### PR TITLE
agate: support memtable read write

### PR DESCRIPTION
**Pull request by skyzh**
In this PR, we added support for memtable read and write.

Signed-off-by: Alex Chi <iskyzh@gmail.com>

### CODE REVIEW

File: src/memtable.rs [hunk 1]
```diff
@@ -25,31 +27,80 @@ struct MemTableCore {
...
             opt,
-            core: Mutex::new(MemTableCore {
+            core: RwLock::new(MemTableCore {
                 wal,
                 max_version: 0,
             }),
         }
     }
 
     pub fn update_skip_list(&self) -> Result<()> {
```
Comment (TennyZhuang) [inline src/memtable.rs:45]: Please add comment for the function

File: src/memtable.rs [hunk 2]
```diff
@@ -25,31 +27,80 @@ struct MemTableCore {
...
+        let mut max_version = core.max_version;
+        if let Some(ref mut wal) = core.wal {
+            let mut it = wal.iter()?;
+            while let Some(entry) = it.next()? {
+                let ts = get_ts(entry.key);
+                if ts > max_version {
+                    max_version = ts;
+                }
+                let v = Value {
+                    value: Bytes::copy_from_slice(entry.value),
```
Comment (TennyZhuang) [inline src/memtable.rs:56]: Is the clone necessary? Should the bytes be written to skl arena directly?

File: src/memtable.rs [hunk 3]
```diff
@@ -25,31 +27,80 @@ struct MemTableCore {
...
     }
 
-    pub fn put(&self, _key: Bytes, _value: Value) -> Result<()> {
-        unimplemented!()
+    pub fn put(&self, key: Bytes, value: Value) -> Result<()> {
+        let mut core = self.core.write();
+        if let Some(ref mut wal) = core.wal {
+            let entry = Entry {
+                key: key.clone(),
+                value: value.value.clone(),
```
Comment (TennyZhuang) [inline src/memtable.rs:74]: `value.value` is confusing

**Comment by skyzh** [comment]
BTW, the corresponding PR on the develop branch is https://github.com/tikv/agatedb/pull/45. This is a refactor for that, and aims to merge it to master.

File: src/memtable.rs [hunk 4]
```diff
@@ -25,31 +27,80 @@ struct MemTableCore {
...
+        if value.meta & value::VALUE_FIN_TXN != 0 {
+            return Ok(());
+        }
+
+        // write to skiplist
+        let ts = get_ts(&key);
+        self.skl.put(key, value);
+
+        // update max version
+        core.max_version = ts;
```
Comment (BusyJay) [inline src/memtable.rs:93]: What if ts is less than `max_version`?

File: src/memtable.rs [hunk 5]
```diff
@@ -25,31 +27,80 @@ struct MemTableCore {
...
+
+        Ok(())
     }
 
     pub fn sync_wal(&self) -> Result<()> {
-        unimplemented!()
+        let mut core = self.core.write();
```
Comment (BusyJay) [inline src/memtable.rs:99]: Is it possible to get rid of the lock?

**Comment by zhangjinpeng1987** [comment]
@skyzh can we close this PR now?

**Comment by skyzh** [comment]
sure

**Pull request closed without merging**